lms/views/api/exceptions.py

```diff
@@ -194,44 +199,58 @@ def api_error(self):
         """
 
         if hasattr(self.context, "error_code"):
-            return self.error_response(
+            return ErrorResponseBody(
                 error_code=self.context.error_code,
                 details=getattr(self.context, "details", None),
             )
 
         # Exception details are not reported here to avoid leaking internal information.
-        return self.error_response(
-            500,
+        self.request.response.status_int = 500
+        return ErrorResponseBody(
             message=_(
                 "A problem occurred while handling this request. Hypothesis has been notified."
             ),
         )
 
     @forbidden_view_config()
     def forbidden(self):
-        return self.error_response(
-            403, message=_("You're not authorized to view this page.")
-        )
+        self.request.response.status_int = 403
+        return ErrorResponseBody(message=_("You're not authorized to view this page."))
 
     @notfound_view_config()
     def notfound(self):
-        return self.error_response(404, message=_("Endpoint not found."))
+        self.request.response.status_int = 404
+        return ErrorResponseBody(message=_("Endpoint not found."))
+
 
-    def error_response(self, status=400, error_code=None, message=None, details=None):
-        self.request.response.status_int = status
+@dataclass
+class ErrorResponseBody:
+    error_code: str = None
+    message: str = None
+    details: dict = None
+
+    def __json__(self, _request):
+        """
+        Return a JSON-serializable representation of this error response body.
+
+        If a view with renderer="json" returns an ErrorResponseBody object then
+        Pyramid's JSON renderer will call this method to get an object to
+        JSON-serialize for the response body. See:
+        https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/renderers.html#using-a-custom-json-method
+        """
 
-        response = {}
+        body = {}
 
-        if error_code is not None:
-            response["error_code"] = error_code
+        if self.error_code is not None:
+            body["error_code"] = self.error_code
 
-        if message is not None:
-            response["message"] = message
+        if self.message is not None:
+            body["message"] = self.message
 
-        if details is not None:
-            response["details"] = details
+        if self.details is not None:
+            body["details"] = self.details
 
-        return response
+        return body
```

Comment:
Gitlab won't let me make a code suggested ('cos lines have been deleted)... but what about:

```python
return {key: value for key, value in asdict(self).items() if value is not None}
```

This is kind of nice because it's shorter, but it's also nice because it doesn't directly depend on the fields the `dataclass` has, so if they changed this would continue to work.

Reply:
Same issue: it's a nice suggestion for saving space but the next PR would break it. There's gonna be a dataclass field added that doesn't get translated directly into the output dict in the same way